landoapi/api/try_push.py

```diff
@@ -0,0 +1,222 @@
+def email(author: bytes) -> bytes:
+    """Get email of author."""
+    r = author.find(b">")
+    if r == -1:
+        r = None
+    return author[author.find(b"<") + 1 : r]
+
+
+def parse_git_author_information(user_header: bytes) -> tuple[bytes, bytes]:
+    """Parse user's name and email address from a Git style author header.
+
+    Converts a header like 'User Name <user@example.com>' to its separate parts.
+    """
+    return person(user_header), email(user_header)
+
+
+def get_timestamp_from_git_date_header(date_header: bytes) -> str:
+    """Convert a Git patch date header into a timestamp."""
+    header_datetime = dateparse(date_header)
+    return str(math.floor(header_datetime.timestamp()))
+
+
+def get_timestamp_from_hg_date_header(date_header: bytes) -> bytes:
+    """Return the first part of the `hg export` date header.
+
+    >>> get_timestamp_from_hg_date_header(b"1686621879 14400")
+    b"1686621879"
+    """
+    return date_header.split(b" ")[0]
+
+
+def build_revision_from_hgexport_patch(patch: bytes) -> Revision:
+    """Convert an `hg export` formatted `bytes` patch into a `Revision`."""
+    helper = HgPatchHelper(io.BytesIO(patch))
+
+    user = helper.header("User")
+    if not user:
+        raise ValueError("Patch does not have a `User` header.")
+
+    author, email = parse_git_author_information(user)
+
+    date = helper.header("Date")
+    if not date:
+        raise ValueError("Patch does not have a `Date` header.")
+
+    timestamp = get_timestamp_from_hg_date_header(date)
+
+    commit_message = helper.commit_description()
+    if not commit_message:
+        raise ValueError("Patch does not have a commit description.")
+
+    return Revision.new_from_patch(
+        raw_diff=helper.get_diff().decode("utf-8"),
+        patch_data={
+            "author_name": author.decode("utf-8"),
+            "author_email": email.decode("utf-8"),
+            "commit_message": helper.commit_description().decode("utf-8"),
+            "timestamp": timestamp.decode("utf-8"),
+        },
+    )
+
+
+def build_revision_from_git_format_patch(patch: bytes) -> Revision:
+    """Turn a `git format-patch` formatted `bytes` patch into a `Revision`."""
+    helper = GitPatchHelper(io.BytesIO(patch))
+
+    from_header = helper.header(b"From")
+    if not from_header:
+        raise ValueError("Patch does not have a `From:` header.")
+
+    author, email = parse_git_author_information(from_header)
+
+    date = helper.header(b"Date")
+    if not date:
+        raise ValueError("Patch does not have a `Date:` header.")
+
+    timestamp = get_timestamp_from_git_date_header(date)
+
+    return Revision.new_from_patch(
+        raw_diff=helper.get_diff().decode("utf-8"),
+        patch_data={
+            "author_name": author.decode("utf-8"),
+            "author_email": email.decode("utf-8"),
+            "commit_message": helper.commit_description().decode("utf-8"),
+            "timestamp": timestamp,
+        },
+    )
```

Comment:
These two methods are very similar, and can be combined into one method that takes a `helper` argument, and with minor additional changes to accommodate the differences in timestamp/user headers. Just for illustration, e.g., 

```python
def build_revision_from_patch(helper: Helper, patch: bytes) -> Revision:
    """Convert a formatted `bytes` patch into a `Revision`."""
    helper = helper(io.BytesIO(patch))

    user = helper.header(user_header_name)
    if not user:
        raise ValueError("Patch does not have a `User` header.")

    author, email = parse_git_author_information(user)

    date = helper.header(date_header_name)
    if not date:
        raise ValueError("Patch does not have a `Date` header.")

    timestamp = get_timestamp_from_hg_date_header(date)

    commit_message = helper.commit_description()
    if not commit_message:
        raise ValueError("Patch does not have a commit description.")

    return Revision.new_from_patch(
        raw_diff=helper.get_diff().decode("utf-8"),
        patch_data={
            "author_name": author.decode("utf-8"),
            "author_email": email.decode("utf-8"),
            "commit_message": helper.commit_description().decode("utf-8"),
            "timestamp": timestamp.decode("utf-8"),
        },
    )
```